File: NeetCode/09-Backtracking/Subsets.cpp
#include <vector>
using namespace std;
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> empty;
        backtrack(0,empty, result, nums);
        return result;
    }
private:
    void backtrack(int start, vector<int> &temp, vector<vector<int>> &result, vector<int>& nums){
        result.push_back(temp);
        for(int i = start; i < nums.size(); i++){
            temp.push_back(nums[i]);
            backtrack(i+1,temp,result,nums);
            temp.pop_back();
        } 

    }
};

File: NeetCode/09-Backtracking/Subsets.java
import java.util.*;

public class Subsets {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(0,new ArrayList<>(), result, nums);
        return result;
    }
    private void backtrack(int start, List<Integer> temp, List<List<Integer>> result, int[] nums){
        result.add(new ArrayList<>(temp));
        for(int i = start; i < nums.length; i++){
            temp.add(nums[i]);
            backtrack(i+1,temp,result,nums);
            temp.remove(temp.size()-1);
        } 

    }
}
